### ISSUE TITLE
Rule suggestion: error when iterators like map or reduce do not return

### ISSUE THREAD

**Issue by spadgos** [eslint/eslint#1128]
On quite a few occasions, I've been messed up by forgetting to return the memo object in a `reduce` call:

```js
// example: convert ['a', 'b', 'c'] --> {a: 0, b: 1, c: 2}
myArray.reduce(function (memo, item, index) {
  memo[item] = index;
}, {}); // Err: cannot set property 'b' of undefined
```

I would argue that it *never* makes sense to have *no* return statement in:
- `map`
- `reduce`
- `reduceRight`
- `filter`
- `every`
- `some`

...since if you don't have a return statement, you could just use a `forEach` (except for `every`, but that would be a weird construct too).

Also, it *never* makes sense to *have* a return statement in `forEach`. I commonly see the misconception that `forEach` can be short-circuited with `return false`.

For the first list, I would not go so far to ensure that a `return` is guaranteed, since it sometimes is not necessary in each iteration (eg: for `filter`, `every` and `some`, no return value is equivalent to `return false`).

The following would be considered errors/warnings:

```js
arr.map(function (item) {
  someStatements();
});
arr.reduce(function (memo, item) {
  someStatements(function () {
    return somethingInner;
  });
});
arr.forEach(function (item) {
  return true;
});
```

The following would not be considered warnings:

```js
arr.reduceRight(function (memo, item) {
  return memo + item;
});

arr.filter(myFunction); // only function literals checked for simplicity
```

I do see one large potential problem with this, in determining when the iterator functions actually belong to `Array.prototype` and when they're just members on another arbitrary object.

```js
var myMappingObject = {
  map: function (callback) {
    // do something, perhaps with the google maps api?
  }
};
myMappingObject.map(function () {
  someStatements(); // would be incorrectly marked as an error
});
```

What do you think?

**Comment by doberkofler** [comment]
+1

**Comment by gyandeeps** [comment]
Are we missing accepted tag here @nzakas ?

**Comment by IanVS** [comment]
Looks like once https://github.com/eslint/eslint/issues/3530 lands, this could be turned into a very excellent rule.  I know it would have saved my bacon a few times.

**Comment by nzakas** [comment]
Yeah, this was created before we used "accepted". :+1:

**Comment by BYK** [comment]
@IanVS - why do we need #3530 for this? Can't we just check function declaration and expressions that are passed into certain function names?

**Comment by IanVS** [comment]
> in an ideal world we'd have branch tracking checks for return.

Consider:

```js
[1,2,3,4,5].reduce(function (acc, val) {
    if (val % 2 === 0) {
        acc.push(val);
        return acc;
    }
}, []);
```

As far as I can tell, without branch checking, we might look at this and think that we are returning `acc` as needed.  Or maybe I'm missing another solution?

**Comment by ilyavolodin** [comment]
Btw, your example is invalid:-) You should always return from inside `reduce`. And to determine if every code path returns you need #3530

**Comment by IanVS** [comment]
> Btw, your example is invalid

That was exactly the point, it is an example of code this rule should warn about.

**Comment by mysticatea** [comment]
Now, #3530 was merged into master, so I'd like to work on this in next few days.

My plan is that the rule checks every code path in callbacks of array's functional methods returns a value, then the rule will say messages similar to `consistent-return`. I guess the detection way for callbacks of array's functional methods is similar to what `no-invalid-this` is using.

Well, I want to get suggestions of the rule name.
To think the name is too hard to me...

**Comment by ilyavolodin** [comment]
Something like `loop-func-return`? `array-func-return`?

**Comment by nzakas** [comment]
`array-callback-return`?

**Comment by mysticatea** [comment]
@ilyavolodin @nzakas Thank you for suggestions!

I will go with `array-callback-return`.

**Comment by IanVS** [comment]
I'm really excited for this rule, @mysticatea.  Thanks for working on it.

**Comment by joeytwiddle** [comment]
Did the `.forEach()` rule ever get written?

My searches for `no-array-callback-return` and `array-callback-no-return` came up blank.

(One could argue that must-return-something and must-not-return-something are actually the same concern: must-return-appropriate-type)
